src/compiler/scanner.rs

```diff
@@ -0,0 +1,141 @@
+use crate::{CharacterCodes, SyntaxKind};
+
+pub struct Scanner {
+    text: Option<String>,
+    pos: Option<usize>,
+    end: Option<usize>,
+    start_pos: Option<usize>,
+    token_pos: Option<usize>,
+    token: Option<SyntaxKind>,
+}
+
+impl Scanner {
+    pub fn get_text_pos(&self) -> usize {
+        self.pos()
+    }
+
+    pub fn get_token_pos(&self) -> usize {
+        self.token_pos()
+    }
+
+    pub fn scan(&mut self) -> SyntaxKind {
+        self.set_start_pos(self.pos());
+
+        loop {
+            self.set_token_pos(self.pos());
+            if self.pos() >= self.end() {
+                return self.set_token(SyntaxKind::EndOfFileToken);
+            }
+            let ch = code_point_at(self.text(), self.pos());
+
+            match ch {
+                CharacterCodes::asterisk => {
+                    self.set_pos(self.pos() + 1);
+                    return self.set_token(SyntaxKind::AsteriskToken);
+                }
+                CharacterCodes::semicolon => {
+                    self.set_pos(self.pos() + 1);
+                    return self.set_token(SyntaxKind::SemicolonToken);
+                }
+                _ch => {
+                    panic!("Unimplemented");
+                    // let identifier_kind = self.scan_identifier(ch);
+                }
+            }
+        }
+    }
+
+    pub fn set_text(
+        &mut self,
+        new_text: Option<&str>,
+        start: Option<usize>,
+        length: Option<usize>,
+    ) {
+        let text = new_text.unwrap_or("");
+        self.set_text_(text);
+        self.set_end(match length {
+            None => text.len(),
+            Some(length) => start.unwrap() + length,
+        });
+        self.set_text_pos(start.unwrap_or(0));
+    }
+
+    pub fn set_text_pos(&mut self, text_pos: usize) {
+        // Debug_.assert(text_pos >= 0);
```

Comment:
Rust helpfully pointed out that this assertion would never fail with `text_pos` typed as `usize`